### ISSUE TITLE
CloudWatch: Providing Multiple Tags to `resource_arns()` Returns Incorrect Response

### ISSUE THREAD

**Issue by hybby** [grafana/grafana#29499]
**What happened**:
I'm trying to create a "Query" dashboard variable using a CloudWatch data source.  

I want to populate the options using the [resource_arns()](https://grafana.com/docs/grafana/latest/datasources/cloudwatch) function.

I want to provide multiple tags to the `resource_arns()` function (for example, a resource tagged with "environment:production" and "service:service-name"):

```
resource_arns(default,elasticloadbalancing:loadbalancer,[{"Component":["$service"]},{"Environment":["$environment"]}])
```

This function takes "tags" as an argument, which through some [digging](https://github.com/grafana/grafana/pull/14803/files#diff-0c6eff99719e840378467688417315b8d9ce22cddf885432bb1885d8aa624f7eR653) suggests that the [ResourceGroupsTaggingApi:GetResources](https://docs.aws.amazon.com/resourcegroupstagging/latest/APIReference/API_GetResources.html) AWS API is used.

This AWS API should accept a group of tag filters.  When multiple tag filters are provided, the returned resources must satisfy all filters.  From the documentation:

> If you specify more than one filter in a single request, the response returns only those resources that satisfy all specified filters.

However, when I try to provide multiple tag filters, it appears that rather than returning resources which satisfy _both_ filters, I'm getting an excessive list of results.  

I think there is an issue with the way that the tags provided as part of `resource_arns()` are being manipulated to be provided to the AWS API call.  I think that it should be possible to pass a valid --tag-filters object to `resource_arns()` and for it to operate as expected.

**What you expected to happen**:
When I perform a query like this via the **Variables -> Edit** menu on a CloudWatch data source:

```
resource_arns(default,elasticloadbalancing:loadbalancer,[{"Component":["$service"]},{"Environment":["$environment"]}])
```

I should be returned ARNs for resources which have _both_ a tag of "Environment:foo" and "Component:bar".

Instead, I am returned an excessive list of resources which do not match the constraints I have set in the TagFilter.

The AWS API confirms this is possible and returns the correct response, limited to resources that match _both_ constraints (note the slight difference in `--tag-filters` body required):

```
$ environment=foo
$ service=bar

$ aws resourcegroupstaggingapi get-resources --resource-type-filters 'elasticloadbalancing:loadbalancer' --tag-filters '[ { "Key": "Environment", "Values": ["foo"] }, { "Key":"Component", "Values":["bar"] } ]'
{
    "ResourceTagMappingList": [
        {
            "ResourceARN": "arn:aws:elasticloadbalancing:eu-west-1:<omitted>:loadbalancer/net/foo-bar-1234567890ABC/abcd1234abcd1234",
            "Tags": [
                {
                    "Key": "Component",
                    "Value": "bar"
                },
                {
                    "Key": "Environment",
                    "Value": "foo"
                }
            ]
        }
    ]
}
```

I think there is an issue with the way that the tags provided as part of `resource_arns()` are being manipulated to be provided to the AWS API call.  I think that it should be possible to pass a valid --tag-filters object to `resource_arns()` and for it to operate as expected.

**How to reproduce it (as minimally and precisely as possible)**:
Reproduction steps above.

**Anything else we need to know?**:
None

**Environment**:
- Grafana version: Grafana v7.3.3 (2489dc4d3a)
- Data source type & version: CloudWatch (Native, same verson)
- OS Grafana is installed on: CentOS7
- User OS & Browser: MacOSX, Firefox
- Grafana plugins: CloudWatch (Native, same verson)
- Others: None

**Comment by wbrowne** [comment]
Thanks for creating this issue @hybby. I haven't been able to reproduce what you're seeing, unfortunately. Could you clarify what you mean by 
>an excessive list of results.

Do the results share any tags in common? It would be helpful if you could share the request and response for the `/api/tsdb/query` call.

If possible, can you try to reproduce if you try a simple setup, for example, two ec2 instances with shared tag values, and adjusting those to test the expected responses?

**Comment by hybby** [comment]
Thanks for picking up my issue, @wbrowne.  Apologies if it didn't have enough information in it at first.

Could you confirm whether you expect `resource_arns()` to work with multiple tags in the way that I've specified?

To address your initial question:

>     an excessive list of results.

By this, I mean that the `resource_arns()` call seems to be returning every resource of the requested type (`elasticloadbalancing:loadbalancer`) from my account.  It doesn't seem to be applying the tag specification constraints that I've specified.

I've taken your advice and used a CloudFormation template to create two `AWS::ElasticLoadBalancingV2::LoadBalancer` resources with two tags applied (please see the end of this comment for the full template).  This both matches my use-case and requires a bit less config than spinning up EC2 instances 🙂.  I've removed my subnet IDs and VPC IDs for security reasons.

This creates two resources:

- **NetworkLoadBalancer1** (AWS::ElasticLoadBalancingV2::LoadBalancer)
  - Tags
    - Key: Environment, Value: env1
    - Key: Component, Value: service1
- **NetworkLoadBalancer2** (AWS::ElasticLoadBalancingV2::LoadBalancer)
  - Tags
    - Key: Environment, Value: env2
    - Key: Component, Value: service1

This reproduces my use-case: I want to select load balancer ARNs by providing both a value for the "Component" tag _and_ a value for the "Environment" tag.  I want returned resources to have both of these tag values.

To achieve this, I try to create a Query Variable with the following query:

```
resource_arns(default,elasticloadbalancing:loadbalancer,[{"Component":["service1"]},{"Environment":["env1"]}])
```

I expect this to return the ARN for **NetworkLoadBalancer1**.  However, the "Preview of values" field appears to show _all_ `elasticloadbalancing:loadbalancer` resources in my account.  The screenshot below shows what I experience.  Apologies for the red boxes redacting our resource information.

![Screenshot 2020-12-03 at 17 46 09](https://user-images.githubusercontent.com/3418856/101067615-ab920100-358f-11eb-8c7c-d79ca682e186.png)

However, using the AWS ResourceGroupsTaggingAPI directly works as expected and returns only the ARN for **NetworkLoadBalancer1**:

```
$ aws resourcegroupstaggingapi get-resources --resource-type-filters 'elasticloadbalancing:loadbalancer' --tag-filters '[ { "Key": "Environment", "Values": ["env1"] }, { "Key":"Component", "Values":["service1"] } ]'
{
    "ResourceTagMappingList": [
        {
            "ResourceARN": "arn:aws:elasticloadbalancing:eu-west-1:<omitted>:loadbalancer/net/tpoc-Netwo-1FD39UGU9A9Z1/bf5878611b8e8a34",
            "Tags": [
                {
                    "Key": "Component",
                    "Value": "service1"
                },
                {
                    "Key": "Environment",
                    "Value": "env1"
                }
            ]
        }
    ]
}
```

I wasn't able to find any documentation for the `/api/tsdb/query` API call (the closest I could find was https://github.com/grafana/grafana/issues/21046).  I tried to put an API call together, but kept getting a "metric request error", which makes me think I'm not quite formatting it in the correct way for a CloudWatch datasource!

```
$ curl -X POST -H "Authorization: Bearer ${TEST_API_KEY}" -H "Content-Type: application/json" https://<grafana>/api/tsdb/query --data '{ "from": "1607016000", "to": "1607016520", "queries": [ { "datasourceId": 195, "query": "resource_arns(default,elasticloadbalancing:loadbalancer,{\"Component\":[\"service1\"]})" } ] }'
{"message":"Metric request error"}%
```

If you could help me with the API call I'd be more than happy to post the results.  But please note that if it returns a large list of our account's resources, I'd need to redact at least the account ID and resource name for security reasons.
 
I hope this helps - please let me know if I can provide any more information.

## CloudFormation Template Used

```
{
  "AWSTemplateFormatVersion": "2010-09-09",
  "Parameters": {
    "VpcId": {
      "Default": "<omitted>",
      "Type": "String",
      "Description": "The Id of the VPC to attach the environment to"
    },
    "PrivateSubnet1Id": {
      "Default": "<omitted>",
      "Type": "String",
      "Description": "The id of private subnet 1"
    },
    "PrivateSubnet2Id": {
      "Default": "<omitted>",
      "Type": "String",
      "Description": "The id of private subnet 2"
    },
    "PrivateSubnet3Id": {
      "Default": "<omitted>",
      "Type": "String",
      "Description": "The id of private subnet 3"
    }
  },
  "Resources": {
    "NetworkLoadBalancer1": {
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
      "Properties": {
        "Type": "network",
        "Tags": [
          {
            "Value": "env1",
            "Key": "Environment"
          },
          {
            "Value": "service1",
            "Key": "Component"
          }
        ],
        "Scheme": "internal",
        "Subnets": [
          {
            "Ref": "PrivateSubnet1Id"
          },
          {
            "Ref": "PrivateSubnet2Id"
          },
          {
            "Ref": "PrivateSubnet3Id"
          }
        ]
      }
    },
    "NetworkLoadBalancer2": {
      "Type": "AWS::ElasticLoadBalancingV2::LoadBalancer",
      "Properties": {
        "Type": "network",
        "Tags": [
          {
            "Value": "env2",
            "Key": "Environment"
          },
          {
            "Value": "service1",
            "Key": "Component"
          }
        ],
        "Scheme": "internal",
        "Subnets": [
          {
            "Ref": "PrivateSubnet1Id"
          },
          {
            "Ref": "PrivateSubnet2Id"
          },
          {
            "Ref": "PrivateSubnet3Id"
          }
        ]
      }
    }
  },
  "Description": "A template that creates two NLBs with tags for debug purposes"
}
```

**Comment by wbrowne** [comment]
Sorry for the delay @hybby - let me try again based on your example and I'll let you know what I find.

>I'd need to redact at least the account ID and resource name for security reasons.

Absolutely - do everything you need to hide sensitive information.

Regarding the `/api/tsdb/query` call, you should be able to capture the raw request and response information there in your browser (IE right click, `Inspect Element` -> `Network`, and then find the correct request)

**Comment by hybby** [comment]
Ah, excellent, didn't realise I could grab them like that.  Thanks @wbrowne.  I've managed to grab those for you.  The returned ARNs are heavily redacted, but I've tried to show how unique the returned values are.

As an aside, I ran this when I confirmed that I had 47 active load balancers in my account.  The query returned 53 rows.  I presume the difference of 6 may be for load balancers that were created and deleted, but still have metrics present in CloudWatch.

Here's a request payload from the **Dashboard -> Dashboard Settings -> Variables -> $variable -> Edit** screen:

```
{
    "from": "1607944156140",
    "to": "1607965756140",
    "queries": [
        {
            "refId": "metricFindQuery",
            "intervalMs": 1,
            "maxDataPoints": 1,
            "datasourceId": <cloudwatch_datasource>,
            "type": "metricFindQuery",
            "subtype": "resource_arns",
            "region": "<region>",
            "resourceType": "elasticloadbalancing:loadbalancer",
            "tags": [
                {
                    "Component": [
                        "service1"
                    ]
                },
                {
                    "Environment": [
                        "env1"
                    ]
                }
            ]
        }
    ]
}
```

And the response:

```{
    "results": {
        "metricFindQuery": {
            "refId": "metricFindQuery",
            "meta": {
                "rowCount": 53
            },
            "series": null,
            "tables": [
                {
                    "columns": [
                        {
                            "text": "text"
                        },
                        {
                            "text": "value"
                        }
                    ],
                    "rows": [
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/1",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/1"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/2",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/2"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/3",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/3"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/4",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/4"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/5",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/5"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/6",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/6"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/7",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/7"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/8",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/8"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/9",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/9"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/10",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/10"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/11",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/11"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/12",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/12"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/13",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/13"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/14",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/14"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/15",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/15"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/16",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/16"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/17",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/17"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/18",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/18"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/19",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/19"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/20",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/20"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/21",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/21"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/22",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/22"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/23",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/23"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/24",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/24"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/25",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/25"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/26",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/26"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/27",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/app/27"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/28",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/28"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/29",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/29"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/30",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/30"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/31",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/31"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/32",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/32"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/33",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/33"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/34",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/34"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/35",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/35"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/36",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/36"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/37",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/37"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/38",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/38"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/39",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/39"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/40",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/40"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/41",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/41"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/42",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/42"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/43",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/43"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/44",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/44"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/45",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/45"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/46",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/46"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/47",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/47"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/48",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/48"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/49",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/49"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/50",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/50"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/51",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/51"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/52",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/52"
                        ],
                        [
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/53",
                            "arn:aws:elasticloadbalancing:<region>:<account>:loadbalancer/net/53"
                        ]
                    ]
                }
            ],
            "dataframes": null
        }
    }
}
```

**Comment by wbrowne** [comment]
@hybby I think I see the issue.
Instead of 
`resource_arns(default,elasticloadbalancing:loadbalancer,[{"Component":["$service"]},{"Environment":["$environment"]}])`

Try

`resource_arns(default,elasticloadbalancing:loadbalancer,{"Component":["$service"],"Environment":["$environment"]})`

**Comment by hybby** [comment]
Hey @wbrowne - appreciate the quick response.  It seems like that's doing the trick!  

Can't believe I messed up the calling of the function like that - I thought I'd tried all the different possible permutations and combinations, too!

Really glad to know that selecting by multiple tags is possible, though.  Thanks so much for helping with my issue!

**Comment by wbrowne** [comment]
Glad to help @hybby 😄 I think we could do a better job of providing feedback to the user to help debug more easily in this case. Something for us to think about!

**Comment by hybby** [comment]
Thanks @wbrowne - I've added an example query to the documentation as part of https://github.com/grafana/grafana/pull/29838 if you think it would be helpful.  It would have prevented me from raising this issue, so it might be worth adding 🙂
